### ISSUE TITLE
[FEATURE] Built-in Personal Knowledge Base with Semantic RAG

### ISSUE THREAD

**Issue by f1c22000-eng** [anthropics/claude-code#28196]
### Preflight Checklist

- [x] I have searched [existing requests](https://github.com/anthropics/claude-code/issues?q=is%3Aissue%20label%3Aenhancement) and this feature hasn't been requested yet
- [x] This is a single feature request (not multiple features)

### Problem Statement

<html>
<body>
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Summary</h2>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">A built-in personal knowledge management system that lets users curate domain-specific knowledge (gotchas, configs, bug fixes, undocumented behaviors) into structured documents, indexed with semantic search, accessible across Claude.ai, Claude Code, and Claude Mobile — reducing token consumption, hallucination, and repeated context-building across conversations.</p>
<h2 class="text-text-100 mt-3 -mb-1 text-[1.125rem] font-bold">Problem Statement</h2>
<h3 class="text-text-100 mt-2 -mb-1 text-base font-bold">The Knowledge Loss Loop</h3>
<p class="font-claude-response-body break-words whitespace-normal leading-[1.7]">Power users (consultants, senior developers, engineers) accumulate critical knowledge during Claude conversations — workarounds, undocumented API behaviors, deployment gotchas, configuration specifics. This knowledge:</p>
<ol class="[li_&amp;]:mb-0 [li_&amp;]:mt-1 [li_&amp;]:gap-1 [&amp;:not(:last-child)_ul]:pb-1 [&amp;:not(:last-child)_ol]:pb-1 list-decimal flex flex-col gap-1 pl-8 mb-3">
<li class="whitespace-normal break-words pl-2"><strong>Gets lost between conversations</strong> — Memory captures fragments, not structured documentation</li>
<li class="whitespace-normal break-words pl-2"><strong>Has to be re-explained every new chat</strong> — "I'm working on X, using Y, the gotcha is Z..."</li>
<li class="whitespace-normal break-words pl-2"><strong>Wastes tokens on rediscovery</strong> — Claude re-searches the same topics, hits the same wrong answers first</li>
<li class="whitespace-normal break-words pl-2"><strong>Causes repeated hallucinations</strong> — Without curated context, Claude guesses on domain-specific details</li>
</ol>
<h3 class="text-text-100 mt-2 -mb-1 text-base font-bold">Current Workarounds Are Insufficient</h3>
<div class="overflow-x-auto w-full px-2 mb-6">
Current Option | Limitation
-- | --
Memory | Short fragments, no structure, no search
Projects | File uploads, not semantic search. Eats context window
Custom Instructions | 1500 chars, static, no domain organization
MCP + external RAG | Requires self-hosting, no standard approach
Pasting context every chat | Manual, error-prone, wastes tokens

</div>
</body>
</html>

### Proposed Solution

Real Example
I'm a consultant working with 5+ enterprise systems (ERP APIs, manufacturing scheduling software, tax/fiscal schemas, WPF/.NET patterns, warehouse management). Each domain has dozens of gotchas that aren't in official docs. Today, I either re-explain them every conversation or build my own RAG infrastructure from scratch.
Proposed Solution
Personal Knowledge Base (PKB)
A structured, searchable knowledge base where users:

Add knowledge — Paste raw content (error logs, API responses, configs) → Claude formats it into structured documentation automatically
Organize by domain — Hierarchical taxonomy that evolves: domains → subdomains (e.g., bling/api/, dassault/ortems/)
Search semantically — Claude searches PKB before web search for relevant domains, reducing token usage and improving accuracy
Access everywhere — Same knowledge base accessible from Claude.ai, Claude Code (MCP), and mobile
Evolve automatically — System suggests reorganization as knowledge grows (merge domains, create subdomains, archive stale content)

How It Works
User discovers gotcha during conversation
  → Claude detects it's domain-specific knowledge
  → Suggests: "Want to save this to your knowledge base?"
  → User confirms → Claude formats and indexes it
  → Next conversation: Claude finds it via semantic search
  → No re-explanation needed. Fewer tokens. Better answers.
Smart Routing (RAG vs Web Search)
"How to authenticate with Bling API?"
  → PKB has curated doc with gotchas     → Use PKB (score 0.92)
  → Web has generic official docs         → Skip web

"What's new in Python 3.13?"
  → PKB has nothing                       → Use web search
  → Return current info

"Bling API rate limits changed?"
  → PKB has old info (60+ days)           → Use PKB + verify with web
  → Best of both worlds
Auto-Evolving Taxonomy
The most powerful aspect — users don't plan the taxonomy, it evolves:
Month 1: ortems/ (3 docs), apriso/ (2 docs)     ← user creates what makes sense now
Month 2: System detects both are Dassault products → suggests merge into dassault/
Month 3: dassault/ortems/ (8 docs), dassault/apriso/ (5 docs)  ← reorganized automatically
Rules:

Never reorganizes without asking — always suggests, user approves
Detects patterns: common prefixes, shared tags, semantic clusters
Handles multiple scenarios: domains merging into parent, large domains splitting into subdomains, cross-domain themes emerging

### Priority

High - Significant impact on productivity

### Feature Category

MCP server integration

### Use Case Example

Full Vision

Auto-detection: "This seems like a useful gotcha — save to your knowledge base?"
Auto-organization: Evolving taxonomy with merge/split suggestions
Cross-client: Same PKB in web, desktop, mobile, Claude Code
Team sharing: Share domain knowledge bases within organizations
Knowledge quality: Track which entries are most used, suggest updates for stale content

User Profile
This feature would be most valuable for:

Consultants working across multiple client systems
Senior developers maintaining knowledge about complex codebases
DevOps/SRE accumulating runbooks and incident responses
Researchers building domain expertise over time
Any power user on Pro/Max plans who uses Claude daily

### Additional Context

Why This Benefits Anthropic
1. Significant Token Reduction
Instead of Claude spending 500-2000 tokens searching and reasoning about domain-specific questions, a PKB lookup returns the exact answer in ~100 tokens of context. For power users doing 100+ queries/day, this is a massive efficiency gain.
2. Reduced Hallucination
Curated, user-verified knowledge > web search > Claude's training data. PKB provides ground truth for domains where Claude's training data is sparse or outdated.
3. Increased User Retention
Users who build a knowledge base become deeply invested. The more knowledge they curate, the more valuable Claude becomes — a positive flywheel.
4. Competitive Differentiation
No competing AI assistant offers structured, semantic, self-organizing personal knowledge management. This would be a genuine moat.
5. Natural MCP Showcase
This is a perfect demonstration of MCP's value proposition: a Remote MCP server providing semantic search over personal knowledge, accessible from any Claude client.
Technical Architecture (Prototype)
I've built a working prototype that demonstrates feasibility:
Google Drive (structured .md files)
       │
       ▼
Google Cloud Run — FREE tier (MCP Server + ChromaDB)
       │  FastMCP + Streamable HTTP
       │  Scale to zero (zero cost when idle)
       │
       ├─→ Claude.ai (Custom Connector)
       ├─→ Claude Code (MCP remote)
       └─→ Claude Mobile (same connector)

Local sync:
  CLAUDE.md (universal config) syncs via Google Drive
  Tools/packages auto-install on new machines
Key Implementation Details

Storage: Markdown files with structured metadata (domain, subdomain, type, tags)
Chunking: Split by H2 sections (~500 tokens), with overlap
Embeddings: all-MiniLM-L6-v2 (ChromaDB default)
MCP Tools: rkm_search (semantic search), rkm_domains (list domains), rkm_stats (base statistics)
Formatting: Claude CLI formats raw input into structured docs (using subscription, no API cost)
Cost: Zero — Google Cloud Run free tier covers 2M requests/month, typical usage is ~15K/month

What Anthropic Could Build
If this concept were integrated into the platform:
Minimum Viable Feature

"Knowledge Base" section in Claude.ai (like Projects, but with semantic search)
Ability to save conversation snippets as knowledge entries
Claude automatically searches PKB for relevant domains before responding
Accessible from Claude Code via built-in MCP

Full Vision

Auto-detection: "This seems like a useful gotcha — save to your knowledge base?"
Auto-organization: Evolving taxonomy with merge/split suggestions
Cross-client: Same PKB in web, desktop, mobile, Claude Code
Team sharing: Share domain knowledge bases within organizations
Knowledge quality: Track which entries are most used, suggest updates for stale content
